Rewrite multi-layer stage 1 (C API) as net driver dispatcher (#792)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/non-unreal-server-worker-types.md

>**Note:** Your external SpatialOS components must have an ID between 1000 and 2000 to be registered by the pipeline.

> You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` functions. These overloaded functions are paramaterized with a `Worker_ComponentId` and a callback function reference that takes one of the following network operation types as an argument:

Reviewer: Typo in  "paramaterized"

> You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` functions. These overloaded functions are parameterized with a `Worker_ComponentId` and a callback function reference that takes one of the following network operation types as an argument:

Reviewer 2: Strictly it takes a const ref to one of these.

* `Worker_AddComponentOp`
* `Worker_RemoveComponentOp`
* `Worker_AuthorityChangeOp`
* `Worker_ComponentUpdateOp`
* `Worker_CommandRequestOp`
* `Worker_CommandResponseOp`

`Worker_ComponentId` and each network operation type are defined in the [Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/api-reference).

> You must register the callbacks inside your game instance's `::Init()` function to ensure your callbacks will be triggered for initial network operations received by the SpatialOS worker connection.

Reviewer 3: nit: This reads like you *must* do it for all callbacks, when really it's only if you might receive ops on initial connection.

> To ensure your callbacks will be triggered for initial network operations received by the SpatialOS worker connection, you will need to register the callbacks inside your game instance's `OnConnected` event callback.

Reviewer 4: This is a super long sentence.
Suggest breaking it up and losing the predictive future (will) passive as it's a bit of a mind bend to read.
```suggestion
You need to ensure that the SpatialOS worker connection receives your callbacks initial network operations. To do this,  register the callbacks inside your game instance's `OnConnected` event callback.
```
Assuming I have understood this correctly. :-)

Author: More of a "you'll likely want to ensure", how to phrase that?

Reviewer 4: If you want to ensure that the SpatialOS worker connection receives your callback for initial network operations, you need to register the callbacks inside your game instance's `OnConnected` event callback.

Reviewer 4: Suggestion - it's one sentence (bad) but it's a bit easier to follow than the original, I think (good).

Each `USpatialDispatcher::AddOpCallback` function returns a `CallbackId`. You can deregister your callbacks using the `USpatialDispatcher::RemoveOpCallback` function and passing the `CallbackId` parameter that was returned by the corresponding call to `USpatialDispatcher::AddOpCallback`.

> There is a basic example in the _Examples_ section below. For more examples of how to deserialize the `Worker_Op` type see the SpatialOS documentation on [serialization in the Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/serialization).

Reviewer: I've read the phrase "Worker SDK in C’s API" a few times now and it still feels weird. I understand the structure and what it means but could we perhaps replace it with something simpler like "C Worker SDK", "Worker SDK C API" or "C API of the Worker SDK"?

Author: I would agree, except this was Laura's wording so I'm inclined to stick with it

```

> Worker_RequestId SendSomeCommandResponse(Worker_EntityId TargetEntityId, Worker_ComponentId ComponentId, Schema_FieldId CommandId) {

Reviewer: In this example could you perhaps show something cleaner using `AddStringToSchema`. Not a big deal, but it will probably help someone trying to write such code discover the `SchemaUtils` and see more examples of how to work with schema objects.

Author: Hmm, schema utils looks good but it's a pretty limited subset of functionality that I'd be scared to see change and accidentally break the docs. I'd prefer to play it safer for now and just use the direct C API calls. Could reference the SchemaUtils file in the doc?

#### Receive data
You could receive and deserialize a component update and command request in your Unreal project code in the following way:
```
void UTPSGameInstance::Init()
{
    // OnConnected is an event declared in USpatialGameInstance

>     OnConnected.AddLambda([this]() {

Reviewer 2: This seems confusing given the previous instructions. We've told users to register callbacks in Init but then here we are deferring all the callbacks to some later time. The lambda is defined in Init but none of the callbacks are necessarily registered there.
However it would seem that making callbacks in OnConnected should be fine, so we should possibly just tell users to do that.

        // On the client the world may not be completely set up, if s we can use the PendingNetGame
        USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver());
        if (NetDriver == nullptr)
        {

>             NetDriver = Cast<USpatialNetDriver>(GetWorldContext()->PendingNetGame->GetNetDriver());

Reviewer: Could this also be null?

Author: In a number of places all over the GDK we do this check and depend on it, so I'd say hesitantly that no it won't be?

```
UCLASS()
class UTestEntitySnapshotGeneration : public USnapshotGenerationTemplate
{

> 	GENERATED_BODY()

Reviewer: This and the newly added code below is using tabs while the snippet above looks like it's using spaces. Not sure if this could affect rendering?

Author: Mmm I just copied both from VS so not sure how this happened. Looks identical in improbadoc preview though.

		// Serialize FromUnreal component data
		Worker_ComponentData UnrealAuthoritativeComponentData{};
		UnrealAuthoritativeComponentData.component_id = 1338;
		UnrealAuthoritativeComponentData.schema_type = Schema_CreateComponentData(1338);
		Schema_Object* UnrealAuthoritativeComponentDataObject = Schema_GetComponentDataFields(UnrealAuthoritativeComponentData.schema_type);
		Schema_AddInt32(UnrealAuthoritativeComponentDataObject, 1, 1); // set other_counter field to 1 initially

		Components.Add(improbable::Position(improbable::Origin).CreatePositionData());
		Components.Add(improbable::Metadata(TEXT("TestEntity")).CreateMetadataData());
		Components.Add(improbable::Persistence().CreatePersistenceData());
		Components.Add(improbable::EntityAcl(AnyWorkerPermission, ComponentWriteAcl).CreateEntityAclData());
		Components.Add(NonUnrealAuthoritativeComponentData);
		Components.Add(UnrealAuthoritativeComponentData);

		TestEntity.component_count = Components.Num();
		TestEntity.components = Components.GetData();

		bool bSuccess = Worker_SnapshotOutputStream_WriteEntity(OutputStream, &TestEntity) != 0;

> 		if (bSuccess) {

Reviewer 3: ```suggestion
		if (bSuccess)
		{
```
